### PR TITLE
SLO - SLO test start

### PR DESCRIPTION
**Pull request by blen-desta**
This PR is a continuation of https://github.com/connexta/saml-conformance/pull/148

#### What does this PR do (if it's not clear from the Title)? Please include any specification references that apply.
- Removed SOAP and PAOS from test's sp metadata and changed second sp port to 8994
- Started post and redirect tests for SLO (sinfle and multiple SP logouts)
- Removed the `--implementation`, `--debug` and `--linent` options from Command.kt and the README since it wasn't working
- Renamed getServiceProvider to getImplementation because Service Provider is overloaded

To test this PR the changes from https://github.com/codice/ddf/pull/3273 are required.

#### Checklist:
- [ ] SAML Spec Table of Contents documentation updated
- [ ] Unit Tests Added/Modified

### CODE REVIEW

File: ctk/idp/src/main/kotlin/org/codice/compliance/web/slo/RedirectSLOTest.kt [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                isSamlRequest = true
+            }.decodeAndVerify()
+            CoreLogoutRequestProtocolVerifier(samlLogoutRequestDom).verify()
+
+            val secondSPLogoutResponse = createDefaultLogoutResponse(samlLogoutRequestDom, true)
+            val encodedSecondSPLogoutResponse = encodeRedirectRequest(secondSPLogoutResponse)
+            val secondSPResponseQueryParams = SimpleSign().signUriString(
+                SAML_RESPONSE,
+                encodedSecondSPLogoutResponse,
+                null)
```
Comment (bakejeyner) [inline ctk/idp/src/main/kotlin/org/codice/compliance/web/slo/RedirectSLOTest.kt]: ❓ Should this relay state parameter be null? In the corresponding POST logout test, it isn't https://github.com/connexta/saml-conformance/pull/151/files#diff-2699a8d35a33214ad873f4e892322231R105.

File: ctk/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt [hunk 2]
```diff
@@ -134,25 +161,48 @@ class TestCommon {
 
         object UseDSASigningSP : TestCaseExtension {
```
[the remark this answers is not in the document]
Comment (blen-desta) [reply ctk/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt]: This is what @jeter added for the Redirect DSA test where he does `config(extensions = listOf(UseDSASigningSP))`. It switches the sp issuer and entity info to be the DSA SP for that test.

File: ctk/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt [hunk 3]
```diff
@@ -223,51 +275,207 @@ class TestCommon {
...
+                    createDefaultAuthnRequest(binding, DSA_SP_ISSUER, DSA_SP_ENTITY_INFO)
+                }
+
+                return if (binding == HTTP_POST) {
+                    val firstLoginResponse = loginPost(authnRequest)
+                    val finalResponse = getImplementation(IdpSSOResponder::class)
+                        .getResponseForPostRequest(firstLoginResponse)
+                    if (!singleSP) {
+                        useDSAServiceProvider()
+                        loginPost(secondRequest, finalResponse.cookies)
```
[the remark this answers is not in the document]
Comment (blen-desta) [reply ctk/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt]: Yes, we provide the cookies from the first login so we don't get the login page.

File: ctk/idp/src/main/kotlin/org/codice/compliance/web/slo/RedirectSLOTest.kt (hunk 1 above)
Comment (brjeter) [inline ctk/idp/src/main/kotlin/org/codice/compliance/web/slo/RedirectSLOTest.kt]: ❗️ Shouldn't we be getting the relayState from the `secondSPLogoutRequest ` and passing that in here?

Comment (blen-desta) [reply ctk/idp/src/main/kotlin/org/codice/compliance/web/slo/RedirectSLOTest.kt]: My bad that should be `logoutRequestRelayState`

File: ctk/common/src/main/kotlin/org/codice/compliance/utils/RestAssuredExtensions.kt [hunk 4]
```diff
@@ -75,18 +77,20 @@ fun Response.extractSamlResponseForm(): Node? {
                 it.children()
                         .list()
                         .any { formControl ->
-                            SAML_RESPONSE.equals(formControl.getAttribute(NAME),
+                            SAML_RESPONSE.equals(formControl.getAttribute(NAME), ignoreCase = true)
+                                || SAML_REQUEST.equals(formControl.getAttribute(NAME),
```
[the remark this answers is not in the document]
Comment (blen-desta) [reply ctk/common/src/main/kotlin/org/codice/compliance/utils/RestAssuredExtensions.kt]: We are, it's on line 82

File: ctk/common/src/main/kotlin/org/codice/compliance/utils/RestAssuredExtensions.kt [hunk 5]
```diff
@@ -75,18 +77,20 @@ fun Response.extractSamlResponseForm(): Node? {
...
 
 private fun Response.isPostBinding(): Boolean {
-    return this.extractSamlResponseForm() != null
+    return this.extractSamlMessageForm() != null
 }
 
 private fun Response.isRedirectBinding(): Boolean {
-    return this.getHeader("Location")?.contains("$SAML_RESPONSE=") == true
+    return this.getHeader(LOCATION)?.contains("$SAML_RESPONSE=") == true ||
+        this.getHeader(LOCATION)?.contains("$SAML_REQUEST=") == true
```
[the remark this answers is not in the document]
Comment (blen-desta) [reply ctk/common/src/main/kotlin/org/codice/compliance/utils/RestAssuredExtensions.kt]: I'm not sure. I'll double check.

File: ctk/common/src/main/kotlin/org/codice/compliance/verification/binding/BindingVerifier.kt [hunk 6]
```diff
@@ -50,16 +51,23 @@ abstract class BindingVerifier(val httpResponse: Response) {
...
-                    SimpleSign().validateSignature(responseObject.signature)
+                val docElement = node.ownerDocument.documentElement
 
-                responseObject.assertions
-                    .filter { it.isSigned }
-                    .forEach { SimpleSign().validateSignature(it.signature) }
+                val samlResponseObject =
+                    if (node.nodeName.contains("Response"))
+                        OpenSAMLUtil.fromDom(docElement) as StatusResponseType
+                    else OpenSAMLUtil.fromDom(docElement) as RequestAbstractType
```
[the remark this answers is not in the document]
Comment (blen-desta) [reply ctk/common/src/main/kotlin/org/codice/compliance/verification/binding/BindingVerifier.kt]: The problem is this method is in the companion object but `isSamlRequest` isn't.

File: ctk/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt [hunk 7]
```diff
@@ -97,13 +120,17 @@ class TestCommon {
...
 
         private const val DEFAULT_SP_ISSUER = "https://samlhost:8993/services/saml"
-        private const val DSA_SP_ISSUER = "https://samlhostdsa:8993/services/samldsa"
+        private const val DSA_SP_ISSUER = "https://samlhostdsa:8994/services/samldsa"
+
         @JvmField
         var currentSPIssuer = DEFAULT_SP_ISSUER
 
+        // Used to return the relay state in the logout response. It's set by the binding verifiers.
+        var logoutRequestRelayState: String? = null
```
[the remark this answers is not in the document]
Comment (blen-desta) [reply ctk/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt]: `relaySate` is a field in `PostFormVerifier`.

[the remark this answers is not in the document]
Comment (brjeter) [reply ctk/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt]: 😢

[the remark this answers is not in the document]
Comment (blen-desta) [reply ctk/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt]: I honestly don't like this way of doing this but I'm not sure what else we can do.

File: ctk/idp/src/main/kotlin/io/kotlintest/provided/ProjectConfig.kt [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+import com.jayway.restassured.RestAssured
+import com.jayway.restassured.RestAssured.config
+import com.jayway.restassured.config.RedirectConfig.redirectConfig
+import io.kotlintest.AbstractProjectConfig
+
+object ProjectConfig : AbstractProjectConfig() {
+
+    override fun beforeAll() {
+        RestAssured.config = config().redirect(redirectConfig().followRedirects(false))
```
Comment (brjeter) [inline ctk/idp/src/main/kotlin/io/kotlintest/provided/ProjectConfig.kt]: 🎉 I'm so excited to not have to worry about redirects biting us again!

**Comment by blen-desta** [comment]
build now

**Comment by blen-desta** [comment]
Hero'd successfully by @brjeter. Tests passed.

**Pull request merged**